framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -4631,6 +4622,448 @@ private void makeGroundTargetType(
+  private final SimpleAnnotatedTypeScanner<Boolean, TypeVariable> captureScanner =
+      new SimpleAnnotatedTypeScanner<>(
+          (type, other) -> type.getUnderlyingType() == other, Boolean::logicalOr, false);
+
+  /**
+   * Set the annotated bounds for fresh type variable {@code capturedTypeVar}, so that it is the
+   * capture of {@code wildcard}.
+   *
+   * @param typeVarToAnnotatedTypeArg mapping from a (type mirror) type variable to its (annotated
+   *     type mirror) type argument
+   * @param capturedTypeVarToAnnotatedTypeVar mapping from a captured type variable to its {@link
+   *     AnnotatedTypeMirror}
+   * @param wildcard wildcard which is converted to {@code capturedTypeVar}
+   * @param typeVariable type variable for which {@code wildcard} is a type argument
+   * @param capturedTypeVar the fresh type variable which is side-effected by this method
+   */
+  private void annotateCapturedTypeVar(
+      Map<TypeVariable, AnnotatedTypeMirror> typeVarToAnnotatedTypeArg,
+      Map<TypeVariable, AnnotatedTypeVariable> capturedTypeVarToAnnotatedTypeVar,
+      AnnotatedWildcardType wildcard,
+      AnnotatedTypeVariable typeVariable,
+      AnnotatedTypeVariable capturedTypeVar) {
+
+    AnnotatedTypeMirror typeVarUpperBound =
+        typeVarSubstitutor.substituteWithoutCopyingTypeArguments(
+            typeVarToAnnotatedTypeArg, typeVariable.getUpperBound());
+    AnnotatedTypeMirror upperBound =
+        AnnotatedTypes.greatestLowerBound(this, typeVarUpperBound, wildcard.getExtendsBound());
+    capturedTypeVar.setUpperBound(upperBound);
+
+    AnnotatedTypeMirror lowerBound =
+        AnnotatedTypes.leastUpperBound(
+            this, typeVariable.getLowerBound(), wildcard.getSuperBound());
+    capturedTypeVar.setLowerBound(lowerBound);
+
+    // Add as a primary annotation any qualifiers that are the same on the upper and lower bound.
+    AnnotationMirrorSet p =
+        new AnnotationMirrorSet(capturedTypeVar.getUpperBound().getAnnotations());
+    p.retainAll(capturedTypeVar.getLowerBound().getAnnotations());
+    capturedTypeVar.replaceAnnotations(p);
+
+    capturedTypeVarSubstitutor.substitute(capturedTypeVar, capturedTypeVarToAnnotatedTypeVar);
+  }
+
+  /**
+   * Substitutes references to captured type variables.
+   *
+   * <p>Unlike {@link #typeVarSubstitutor}, this class does not copy the type. Call {@code
+   * substitute} to use.
+   */
+  private final CapturedTypeVarSubstitutor capturedTypeVarSubstitutor =
+      new CapturedTypeVarSubstitutor();
+
+  /**
+   * Substitutes references to captured types in {@code type} using {@code
+   * capturedTypeVarToAnnotatedTypeVar}.
+   *
+   * <p>Unlike {@link #typeVarSubstitutor}, this class does not copy the type. Call {@code
+   * substitute} to use.
+   */
+  private static class CapturedTypeVarSubstitutor extends AnnotatedTypeCopier {
+
+    /** A mapping from a captured type variable to an AnnotatedTypeVariable. */
```

Comment:
Which AnnotatedTypeVariable is mapped to?  That is, what is its relationship to the key?